### PR TITLE
AP-86: Add page router

### PR DESCRIPTION
**Pull request by chrisdevereux**

### CODE REVIEW

File: src/lib/components/Link.tsx [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+      if (e.defaultPrevented) {
+        return
+      }
+    }
+
+    if (!isModifiedClick(e)) {
+      return
+    }
+
+    // e.preventDefault()
```
Comment (richard-hulm) [inline src/lib/components/Link.tsx]: Is this meant to be commented out?

File: docs-site/src/pages/docs/getting-started.js [hunk 2]
```diff
@@ -42,7 +42,7 @@ export default () => {
               Pages are picked up from all files under the <code>src/pages</code> directory.
             </p>
             <p>
-              Add a new source file for the app.
+              Add a new source file for the containing a route component for the path <code>/</code>.
```
Comment (robotdevil) [inline docs-site/src/pages/docs/getting-started.js]: This doesn't read quite clearly to me...?

File: src/lib/components/Link.tsx [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    if (this.props.onClick) {
+      this.props.onClick(e)
+
+      if (e.defaultPrevented) {
+        return
+      }
+    }
+
+    if (!isModifiedClick(e)) {
+      return
```
Comment (richard-hulm) [inline src/lib/components/Link.tsx]: What is this intended to do...?

File: src/lib/components/Link.tsx [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    }
+
+    // e.preventDefault()
+  }
+
+  render() {
+    return (
+      <a href={this.props.href} onClick={this.handleClick}>
```
Comment (richard-hulm) [inline src/lib/components/Link.tsx:24]: Will this cause full page refreshes when clicking a link? Or is it doing react-router like redirects?

File: src/lib/core/Router.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+  get route(): Match {
+    const matches = this.recognizer.recognize(this.path)
+    if (!matches) {
+      return undefined
+    }
+
+    const match = matches[0]
+
+    /* istanbul ignore next */
```
Comment (richard-hulm) [inline src/lib/core/Router.ts]: Huh....?

File: src/lib/core/route.test.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+describe('isRouteComponent()', () => {
+  it('should return true for decorated component', () => {
+    class MyRoute extends React.Component { }
+
+    decorateRouteComponent('/foo', MyRoute)
+    expect(isRouteComponent(MyRoute)).to.be.true
+  })
+
+<<<<<<< HEAD
```
Comment (richard-hulm) [inline src/lib/core/route.test.ts]: Conflict issues?

File: src/lib/components/Link.tsx (hunk 1 above)
Comment (robotdevil) [inline src/lib/components/Link.tsx]: Is this comment needed still?

File: src/lib/core/App.tsx [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+  render() {
+    const { route } = this.router
+
+    if (route) {
+      const Handler = route.handler
+      return <Handler {...route} />
+
+    } else {
+      return null
```
Comment (robotdevil) [inline src/lib/core/App.tsx]: What's the result of returning null? Does the user see a 404?

File: src/lib/core/Router.ts (hunk 5 above)
Comment (robotdevil) [inline src/lib/core/Router.ts]: Is this telling the coverage checker to ignore the no match = undefined case - if so, why's that?

File: src/lib/core/route.test.ts (hunk 6 above)
Comment (robotdevil) [inline src/lib/core/route.test.ts]: Looks like there is some merge conflict annotations in here?

Comment (chrisdevereux) [reply src/lib/core/route.test.ts]: Ergh, vscode is really bad at reloading changed files... Will push a fix then rebase it out before merging.

File: src/lib/core/Router.ts (hunk 5 above)
Comment (chrisdevereux) [reply src/lib/core/Router.ts]: The istanbul ignore stuff is for the code coverage tool. Will document why it's being ignored!

File: src/lib/components/Link.tsx (hunk 4 above)
Comment (chrisdevereux) [reply src/lib/components/Link.tsx:24]: Yeah, it's a placeholder component. Will document.

File: src/lib/core/route.test.ts (hunk 6 above)
Comment (chrisdevereux) [reply src/lib/core/route.test.ts]: 👍

File: src/lib/core/Router.ts (hunk 5 above)
Comment (chrisdevereux) [reply src/lib/core/Router.ts]: The istanbul ignore stuff is for the code coverage tool. Will document why it's being ignored!

File: src/lib/core/App.tsx (hunk 7 above)
Comment (chrisdevereux) [reply src/lib/core/App.tsx]: No, it just renders an empty page. Was thinking that 404 should be handled with a route matching '*'

Comment (chrisdevereux) [reply src/lib/core/App.tsx]: Could argue that it should throw an error here instead, though. What do you think?

File: docs-site/src/pages/docs/getting-started.js (hunk 2 above)
Comment (chrisdevereux) [reply docs-site/src/pages/docs/getting-started.js]: 👍

File: src/lib/components/Link.tsx (hunk 1 above)
Comment (chrisdevereux) [reply src/lib/components/Link.tsx]: Placeholder for later. Have documented in new commit.

**Review by richard-hulm — DISMISSED** [review]

File: src/lib/core/App.tsx (hunk 7 above)
Comment (robotdevil) [reply src/lib/core/App.tsx]: I think an error would probably be better - from my reading of the code it looks like an error state. The error would be more descriptive as well...

**Review by robotdevil — APPROVED** [review]

**Pull request merged**